src/utils/codec.ts

```diff
@@ -0,0 +1,318 @@
+                if (point === undefined) {
+                    return '0'.repeat(8)
+                }
+
+                // Complete the 0 prefix from every byte
+                return ('0'.repeat(8) + point.toString(2)).slice(-8)
+            }).reduce((o: string, c: string) => (o += c)),
+        )
+    }
+
+    /**
+     * Convert code point into target radix string
+     *
+     * @name P2RS - Point to Radix String
+     *
+     * @param {CodecScheme} scheme      - target codec scheme
+     * @param {Point}       point       - code point
+     * @param {string}      str         - the output string
+     * @param {boolean}     completion  - if complete the prefix
+     */
+    static P2RS(scheme: CodecScheme, point: Point, str: string, completion: boolean): [number, string] {
+        if (point === undefined) {
+            return [0, '']
+        } else if (point === 0) {
+            return [0, str]
+        }
+
+        // Complete prefix if it is required
+        const reminder = completion
+            ? (scheme.table[0].repeat(scheme.bits - 1) + scheme.table[point & ((1 << scheme.bits) - 1)]).slice(
+                  -scheme.bits,
+              )
+            : scheme.table[point & ((1 << scheme.bits) - 1)]
+
+        // Recursive result
+        return this.P2RS(scheme, point >> scheme.bits, reminder + str, completion)
+    }
+
+    /**
+     * Convert code points into target radix string
+     *
+     * @name Ps2RS - Points to Radix String
+     *
+     * @param {CodecScheme} scheme     - target codec scheme.
+     * @param {Point[]}     points     - code points
+     * @param {boolean}     completion - complete prefix
+     */
+    static Ps2RS(scheme: CodecScheme, points: Point[], completion: boolean): string {
+        return points.map((point: Point) => Codec.P2RS(scheme, point, '', completion)[1]).reduce((o, c) => (o += c))
+    }
+
+    /**
+     * This binary is converted from 8 bits points
+     */
+    private bin: string
+
+    /**
+     * Build Codec from binary string
+     *
+     * @param {string} binary - binary string
+     */
+    constructor(binary: string) {
+        this.bin = binary
+    }
+
+    /**
+     * Convert points to hex
+     */
+    public toBinary(): string {
+        return this.bin
+    }
+
+    /**
+     * Convert points to hex
+     */
+    public toHex(): string {
+        return this.toSchemeString(HEX)
+    }
+
+    /**
+     * Convert points to hex
+     */
+    public toBase64(): string {
+        return this.toSchemeString(BASE64)
+    }
+
+    /**
+     * Convert codec binary to target scheme string
+     *
+     * @param {CodecScheme} scheme - target codec scheme
+     *
+     * @public This method can be used by manual `CodecSchema`.
+     *
+     * @return {string} schema string
+     */
+    // prettier-ignore
+    public toSchemeString(scheme: CodecScheme): string {
+        const bytes = this.bin.match(new RegExp(`.{1,${scheme.bits}}`, "g"));
```

Comment:
plz check schema.bits is corrent.

```typescript
const schema = { bits: NaN }
codec. toSchemeString(schema) // BOOM
```

Reply:
XDDD

Reply:
I'm hesitating about throwing `Error` or return an empty string?

Reply:
throwing Error, plz use the https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#assertion-functions feature

```
assertCodeScheme(scheme)
```